File: js-src/pieceView.js
/**
 * Handles the visual representation of a Tetris piece.
 * Converted from Java PieceView class.
 */
export class PieceView {
  /**
   * Creates a new PieceView instance.
   * @param {string} symbol - The character symbol to represent this piece
   */
  constructor(symbol) {
    this.symbol = symbol;
  }

  /**
   * Gets the piece symbol.
   * @returns {string} The piece symbol
   */
  getSymbol() {
    return this.symbol;
  }

  /**
   * Sets the piece symbol.
   * @param {string} symbol - The new piece symbol
   */
  setSymbol(symbol) {
    this.symbol = symbol;
  }

  /**
   * Renders the piece onto the display grid.
   * @param {Piece} piece - The piece to render
   * @param {string[][]} displayGrid - The grid to render onto
   */
  render(piece, displayGrid) {
    const shape = piece.getShape();
    const pos = piece.getPosition();

    for (let i = 0; i < shape.length; i++) {
      for (let j = 0; j < shape[i].length; j++) {
        if (shape[i][j]) {
          const x = pos.getX() + j;
          const y = pos.getY() + i;
          if (y >= 0 && y < displayGrid.length && x >= 0 && x < displayGrid[0].length) {
            displayGrid[y][x] = this.symbol;
          }
        }
      }
    }
  }
}

File: cpp_tetris/src/PieceView.cpp
#include "tetris/PieceView.h"

namespace tetris {

PieceView::PieceView(char symbol) : symbol_(symbol) {}

char PieceView::getSymbol() const {
    return symbol_;
}

void PieceView::setSymbol(char symbol) {
    symbol_ = symbol;
}

void PieceView::render(const Piece& piece, std::vector<std::vector<char>>& displayGrid) const {
    const auto& shape = piece.getShape();
    const auto& pos = piece.getPosition();

    for (int i = 0; i < static_cast<int>(shape.size()); ++i) {
        for (int j = 0; j < static_cast<int>(shape[i].size()); ++j) {
            if (shape[i][j]) {
                int x = pos.getX() + j;
                int y = pos.getY() + i;
                
                if (y >= 0 && y < static_cast<int>(displayGrid.size()) && 
                    x >= 0 && x < static_cast<int>(displayGrid[0].size())) {
                    displayGrid[y][x] = symbol_;
                }
            }
        }
    }
}

} // namespace tetris